src/lib.rs

```diff
@@ -154,65 +157,122 @@ mod platform {
-        // counter has reached its maximum value or if the semaphore
-        // is invalid, we can therefore safely ignore return value.
-        ReleaseSemaphore(SEMAPHORE, 1, ptr::null_mut());
+        // Assuming this always succeeds. Can't really handle errors in any meaningful way.
+        self::kernel32::ReleaseSemaphore(SEMAPHORE, 1, ptr::null_mut());
         TRUE
     }
 
-    /// Register os signal handler, must be called before calling block_ctrl_c().
-    /// Should only be called once.
+    /// Register os signal handler.
+    ///
+    /// Must be called before calling [`block_ctrl_c()`](fn.block_ctrl_c.html)
+    /// and should only be called once.
+    ///
+    /// # Errors
+    /// Will return an error if a recoverable system error occur. An error should leave
+    /// the system in a constant state, where trying again might succeed.
+    ///
+    /// # Panics
+    /// Will panic if a non recoverable system error occur.
+    ///
     #[inline]
     pub unsafe fn init_os_handler() -> Result<(), Error> {
-        SEMAPHORE = CreateSemaphoreA(ptr::null_mut(), 0, MAX_SEM_COUNT, ptr::null());
+        SEMAPHORE = self::kernel32::CreateSemaphoreA(ptr::null_mut(), 0, MAX_SEM_COUNT, ptr::null());
         if SEMAPHORE.is_null() {
-            return Err(Error::Init("CreateSemaphoreA failed".into()));
+            return Err(Error::System(io::Error::last_os_error()));
         }
-        if SetConsoleCtrlHandler(Some(os_handler), TRUE) == FALSE {
-            return Err(Error::SetHandler);
+
+        if self::kernel32::SetConsoleCtrlHandler(Some(os_handler), TRUE) == FALSE {
+            let e = io::Error::last_os_error();
+            self::kernel32::CloseHandle(SEMAPHORE);
+            SEMAPHORE = 0 as HANDLE;
+            return Err(Error::System(e));
         }
+
         Ok(())
     }
 
     /// Blocks until a Ctrl-C signal is received.
+    ///
+    /// Must be called after calling [`init_os_handler()`](fn.init_os_handler.html).
+    ///
+    /// # Errors
+    /// Will return an error if a system error occurs.
+    ///
     #[inline]
-    pub unsafe fn block_ctrl_c() {
-        assert_eq!(WaitForSingleObject(SEMAPHORE, INFINITE), WAIT_OBJECT_0);
+    pub unsafe fn block_ctrl_c() -> Result<(), Error> {
+        use self::winapi::{INFINITE, WAIT_OBJECT_0, WAIT_FAILED};
+
+        match self::kernel32::WaitForSingleObject(SEMAPHORE, INFINITE) {
+            WAIT_OBJECT_0 => Ok(()),
+            WAIT_FAILED => Err(Error::System(io::Error::last_os_error())),
+            ret => Err(Error::System(io::Error::new(
+                io::ErrorKind::Other,
+                format!("WaitForSingleObject(), unexpected return value \"{:x}\"", ret),
+            ))),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn raise_ctrl_c() {
+        unsafe {
+            // This will signal the whole process group.
+            assert!(self::kernel32::GenerateConsoleCtrlEvent(self::winapi::CTRL_C_EVENT, 0) != 0);
```

Comment:
Yea, I am looking at workaround where we detach from current console and open our own while we run our tests, but I am having trouble getting stdout to work in the new console.